presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantEnforceSingleRowNode.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.enforceSingleRow;
+
+/**
+ * Remove EnforceSingleRow node when the subplan is guaranteed to produce only one row
+ */
+public class RemoveRedundantEnforceSingleRowNode
+        implements Rule<EnforceSingleRowNode>
+{
+    private static final Pattern<EnforceSingleRowNode> PATTERN = enforceSingleRow()
+            .matching(RemoveRedundantEnforceSingleRowNode::isAtMost);
+
+    @Override
+    public Pattern<EnforceSingleRowNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    private static boolean isAtMost(EnforceSingleRowNode node)
```

Comment:
call this isAtMostOne()

Reply:
Turns out that null comparisons require the EnforceSingleRowNode in the plan. I have changed it to be removed only with the cardinality of the source is provably one. isAtMostOne -> isSourceScalar

Reply:
Looks like the null semantics are a little more complicated that I initially thought. I will look into this more deeply.